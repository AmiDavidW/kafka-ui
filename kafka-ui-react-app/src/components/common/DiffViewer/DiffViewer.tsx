/* eslint-disable react/jsx-props-no-spreading */
import { diff as DiffEditor } from 'react-ace';
import 'ace-builds/src-noconflict/mode-json5';
import 'ace-builds/src-noconflict/theme-textmate';
import React from 'react';
import { IDiffEditorProps } from 'react-ace/lib/diff';

interface DiffViewerProps extends IDiffEditorProps {
  isFixedHeight?: boolean;
}

const DiffViewer = React.forwardRef<DiffEditor | null, DiffViewerProps>(
  (props, ref) => {
    const { isFixedHeight, ...rest } = props;
    const autoHeight =
      !isFixedHeight && props.value && props.value.length === 2
        ? Math.max(
            props.value[0].split(/\r\n|\r|\n/).length + 1,
            props.value[1].split(/\r\n|\r|\n/).length + 1
          ) * 16
        : 500;
    console.log(rest);
    console.log(props);

    return (
      <DiffEditor
        name="diff-editor"
        ref={ref}
        mode="json5"
        theme="textmate"
        tabSize={2}
        width="100%"
        height={`${autoHeight}px`}
        showPrintMargin={false}
        maxLines={Infinity}
        readOnly
        wrapEnabled
        {...rest}
      />
    );
  }
);

DiffViewer.displayName = 'DiffViewer';

export default DiffViewer;
