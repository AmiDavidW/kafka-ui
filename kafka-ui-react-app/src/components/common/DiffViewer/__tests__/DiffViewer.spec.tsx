import { shallow } from 'enzyme';
import React from 'react';
import DiffViewer from 'components/common/DiffViewer/DiffViewer';

describe('Editor component', () => {
  const left = '{\n}';
  const right = '{\ntest: true\n}';
  it('matches the snapshot', () => {
    const component = shallow(<DiffViewer value={[left, right]} name="name" />);
    expect(component).toMatchSnapshot();
  });

  it('matches the snapshot with fixed height', () => {
    const component = shallow(
      <DiffViewer value={[left, right]} name="name" isFixedHeight />
    );
    expect(component).toMatchSnapshot();
  });

  it('matches the snapshot with fixed height with no value', () => {
    const component = shallow(<DiffViewer name="name" isFixedHeight />);
    expect(component).toMatchSnapshot();
  });

  it('matches the snapshot without fixed height with no value', () => {
    const component = shallow(<DiffViewer name="name" />);
    expect(component).toMatchSnapshot();
  });

  it('matches the snapshot without fixed height with one value', () => {
    const component = shallow(<DiffViewer value={[left]} name="name" />);
    expect(component).toMatchSnapshot();
  });
});
