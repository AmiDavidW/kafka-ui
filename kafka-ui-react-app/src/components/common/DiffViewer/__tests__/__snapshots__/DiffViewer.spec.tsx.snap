// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor component matches the snapshot 1`] = `
<DiffComponent
  cursorStart={1}
  editorProps={Object {}}
  enableBasicAutocompletion={false}
  enableLiveAutocompletion={false}
  focus={false}
  fontSize={12}
  height="64px"
  highlightActiveLine={true}
  maxLines={Infinity}
  minLines={null}
  mode="json5"
  name="name"
  onChange={null}
  onLoad={null}
  onPaste={null}
  onScroll={null}
  orientation="beside"
  readOnly={true}
  scrollMargin={
    Array [
      0,
      0,
      0,
      0,
    ]
  }
  setOptions={Object {}}
  showGutter={true}
  showPrintMargin={false}
  splits={2}
  style={Object {}}
  tabSize={2}
  theme="textmate"
  value={
    Array [
      "{
}",
      "{
test: true
}",
    ]
  }
  width="100%"
  wrapEnabled={true}
/>
`;

exports[`Editor component matches the snapshot with fixed height 1`] = `
<DiffComponent
  cursorStart={1}
  editorProps={Object {}}
  enableBasicAutocompletion={false}
  enableLiveAutocompletion={false}
  focus={false}
  fontSize={12}
  height="500px"
  highlightActiveLine={true}
  maxLines={Infinity}
  minLines={null}
  mode="json5"
  name="name"
  onChange={null}
  onLoad={null}
  onPaste={null}
  onScroll={null}
  orientation="beside"
  readOnly={true}
  scrollMargin={
    Array [
      0,
      0,
      0,
      0,
    ]
  }
  setOptions={Object {}}
  showGutter={true}
  showPrintMargin={false}
  splits={2}
  style={Object {}}
  tabSize={2}
  theme="textmate"
  value={
    Array [
      "{
}",
      "{
test: true
}",
    ]
  }
  width="100%"
  wrapEnabled={true}
/>
`;

exports[`Editor component matches the snapshot with fixed height with no value 1`] = `
<DiffComponent
  cursorStart={1}
  editorProps={Object {}}
  enableBasicAutocompletion={false}
  enableLiveAutocompletion={false}
  focus={false}
  fontSize={12}
  height="500px"
  highlightActiveLine={true}
  maxLines={Infinity}
  minLines={null}
  mode="json5"
  name="name"
  onChange={null}
  onLoad={null}
  onPaste={null}
  onScroll={null}
  orientation="beside"
  readOnly={true}
  scrollMargin={
    Array [
      0,
      0,
      0,
      0,
    ]
  }
  setOptions={Object {}}
  showGutter={true}
  showPrintMargin={false}
  splits={2}
  style={Object {}}
  tabSize={2}
  theme="textmate"
  value={
    Array [
      "",
      "",
    ]
  }
  width="100%"
  wrapEnabled={true}
/>
`;

exports[`Editor component matches the snapshot without fixed height with no value 1`] = `
<DiffComponent
  cursorStart={1}
  editorProps={Object {}}
  enableBasicAutocompletion={false}
  enableLiveAutocompletion={false}
  focus={false}
  fontSize={12}
  height="500px"
  highlightActiveLine={true}
  maxLines={Infinity}
  minLines={null}
  mode="json5"
  name="name"
  onChange={null}
  onLoad={null}
  onPaste={null}
  onScroll={null}
  orientation="beside"
  readOnly={true}
  scrollMargin={
    Array [
      0,
      0,
      0,
      0,
    ]
  }
  setOptions={Object {}}
  showGutter={true}
  showPrintMargin={false}
  splits={2}
  style={Object {}}
  tabSize={2}
  theme="textmate"
  value={
    Array [
      "",
      "",
    ]
  }
  width="100%"
  wrapEnabled={true}
/>
`;

exports[`Editor component matches the snapshot without fixed height with one value 1`] = `
<DiffComponent
  cursorStart={1}
  editorProps={Object {}}
  enableBasicAutocompletion={false}
  enableLiveAutocompletion={false}
  focus={false}
  fontSize={12}
  height="500px"
  highlightActiveLine={true}
  maxLines={Infinity}
  minLines={null}
  mode="json5"
  name="name"
  onChange={null}
  onLoad={null}
  onPaste={null}
  onScroll={null}
  orientation="beside"
  readOnly={true}
  scrollMargin={
    Array [
      0,
      0,
      0,
      0,
    ]
  }
  setOptions={Object {}}
  showGutter={true}
  showPrintMargin={false}
  splits={2}
  style={Object {}}
  tabSize={2}
  theme="textmate"
  value={
    Array [
      "{
}",
    ]
  }
  width="100%"
  wrapEnabled={true}
/>
`;
